File: 28.implement-str-str/solution.py
from typing import List


class Solution:
    def buildLSP(self, s: str) -> List[int]:
        lsp = [0] * len(s)
        
        i, j = 1, 0
        
        while i < len(s):
            if s[i] == s[j]:
                lsp[i] = j + 1
                i += 1
                j += 1
            elif j == 0:
                i += 1
            else:
                j = lsp[j-1]
                
        
        return lsp

    def strStr(self, haystack: str, needle: str) -> int:
        n = len(haystack)
        m = len(needle)
        
        i = 0
        j = 0
        
        lps = self.buildLSP(needle)
        
        while i < n and j < m:

            
            if haystack[i] == needle[j]:
                i += 1
                j += 1
                
                if j == m:
                    break
        
            elif j > 0:
                j = lps[j-1]
            else:
                i += 1
                
        return i - j if j == m else -1

s = Solution()
s.strStr('hello', 'll')

File: 28.implement-str-str/solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int rec(string haystack, string needle, int left, int right, unordered_map<string, int>& memo) {
        string sub = haystack.substr(left, right-left+1);
        if (sub == needle) {
            return left;
        }

        if (left >= right) {
            return -1;
        }

        if (memo.find(sub) != memo.end()) {
            return memo[sub];
        }

        int option1 = this->rec(haystack, needle, left, right-1, memo);
        int option2 = this->rec(haystack, needle, left+1, right, memo);

        int ans = option1 != -1 ? option1 : option2;
        memo[sub] = ans;
        return memo[sub];
    }

    int strStr(string haystack, string needle) {
        if (needle.size() == 0) {
            return 0;
        }
        
        if (haystack == needle) {
            return 0;
        }

        int left = 0;
        int right = haystack.size() - 1;

        unordered_map<string, int> memo;

        int option1 = this->rec(haystack, needle, left, right-1, memo);
        int option2 = this->rec(haystack, needle, left+1, right, memo);

        return option1 != -1 ? option1 : option2;    
    }
};

int main(int argc, char const *argv[])
{
    Solution s;
    cout << s.strStr("abbbbbaabbaabaabbbaaaaabbabbbabbbbbaababaabbaabbbbbababaababbbbaaabbbbabaabaaaabbbbabbbaabbbaabbaaabaabaaaaaaaa", "abbbaababbbabbbabbbbbabaaaaaaabaabaabbbbaabab") << endl;
    return 0;
}
